### PR TITLE
Use containers labels instead of task labels from Docker Swarm SD (#9187)

### LINKED ISSUE #9187: Docker Swarm SD does not add container labels as __meta_dockerswarm_task_label_<labelname>
**What did you do?**
Based loosely on the [dockerswarm SD documentation](https://prometheus.io/docs/guides/dockerswarm/#monitoring-containers), I tried to set up service discovery with a small tweak: using _task labels_ rather than _service labels_.

Example docker-stack.yml:
```yaml
version: "3.5"

services:
  example:
    image: example
    labels:
      prometheus-job: containerlabel
    deploy:
      replicas: 2
```

And the associated `prometheus.yml` -- this is directly copied from [Prometheus documentation](https://prometheus.io/docs/guides/dockerswarm/#monitoring-containers) but uses "\_\_meta\_dockerswarm\_**task**\_label\_" rather than "\_\_meta\_dockerswarm\_**service**\_label\_"
```yaml
scrape_configs:
  - job_name: 'dockerswarm'
    dockerswarm_sd_configs:
      - host: unix:///var/run/docker.sock
        role: tasks
    relabel_configs:
      # Only keep containers that should be running.
      - source_labels: [__meta_dockerswarm_task_desired_state]
        regex: running
        action: keep
      # Only keep containers that have a `prometheus-job` label.
      - source_labels: [__meta_dockerswarm_task_label_prometheus_job]
        regex: .+
        action: keep
      # Use the prometheus-job Swarm label as Prometheus job label.
      - source_labels: [__meta_dockerswarm_task_label_prometheus_job]
        target_label: job
```

**What did you expect to see?**
I expected Prometheus to identify each task instance as a target through dockerswarm service discovery, but the `__meta_dockerswarm_task_label_*` meta-labels don't seem to be populated:
image

The labels _do_ seem to be set correctly -- running `docker service inspect example` gives something like:
```
// ID, Version, etc...
"Spec": {
    "Name": "example_example",
    "Labels": {
        "com.docker.stack.image": "example",
        "com.docker.stack.namespace": "example",
    },
    "TaskTemplate": {
        "ContainerSpec": {
            "Image": "example:latest",
            "Labels": {
                "com.docker.stack.namespace": "localization",
                "prometheus-job": "containerlabel",       // <---- label is correctly set!
            },
// blah blah...
}
```

Poking around in the docker swarm SD source shows we're populating task labels by querying the docker client API for a TaskList:

https://github.com/prometheus/prometheus/blob/39d79c3cfb86c47d6bc06a9e9317af582f1833bb/discovery/moby/tasks.go#L46

then, for each task in the returned values, adding `task.Labels`:

https://github.com/prometheus/prometheus/blob/39d79c3cfb86c47d6bc06a9e9317af582f1833bb/discovery/moby/tasks.go#L78-L81

However, as far as I can tell, the container labels are not set on `task.Labels` but instead on `task.Spec.ContainerSpec.Labels`: a cheeky little test with the docker client seems to support this.

<details>
  <summary>Example go source -- click to expand</summary>
  
```go
package main

import (
	"context"
	"fmt"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/client"
)

func main() {
	opts := []client.Opt{
		client.WithHost("unix:///var/run/docker.sock"),
		client.WithAPIVersionNegotiation(),
	}

	c, err := client.NewClientWithOpts(opts...)

	if err != nil {
		fmt.Errorf("Failed to create client, %w", err)
	}

	ctx := context.TODO()
	tasks, err := c.TaskList(ctx, types.TaskListOptions{})

	if err != nil {
		fmt.Errorf("Failed to query task list, %w", err)
	}

	for _, t := range tasks {
		fmt.Printf("task.Labels is empty: %s\n", t.Labels)
		fmt.Printf("but task.Spec.ContainerSpec.Labels is not: %s\n", t.Spec.ContainerSpec.Labels)
	}
}
```
</details>

gives me:

```
task.Labels is empty: map[]
but task.Spec.ContainerSpec.Labels is not: map[com.docker.stack.namespace:example, prometheus-job:containerlabel]
```

This was somewhat touched on in the discussion for https://github.com/prometheus/prometheus/issues/8272#issuecomment-857699769 but I'm still unclear on how to work with "task labels".

Are task labels == container labels? If not, how can we set task labels in docker swarm that would be picked up by Prometheus's service discovery? Otherwise, should https://github.com/prometheus/prometheus/blob/39d79c3cfb86c47d6bc06a9e9317af582f1833bb/discovery/moby/tasks.go#L78 use `s.Spec.ContainerSpec.Labels` rather than `s.Labels`?

### PR DESCRIPTION
**Pull request by importhuman**
Signed-off-by: Ujjwal Goyal <importujjwal@gmail.com>

Closes #9187. 

Please let me know if any modifications are needed, either in the code or elsewhere.

### CODE REVIEW

**Comment by roidelapluie** [comment]
can we use containers labels instead? Thanks!

**Comment by importhuman** [comment]
@roidelapluie Did you mean adding a label similar to what is in [discovery/moby/docker.go](https://github.com/prometheus/prometheus/blob/main/discovery/moby/docker.go#L202-L205)? If the latest commit is along the correct lines, are any changes required? I presume we can add some tests based on the ones in `tasks_test.go` and `docker_test.go`?

**Comment by importhuman** [comment]
@roidelapluie Hi, can you please let me know what changes are required, if any?

**Comment by roidelapluie** [comment]
I think the current version renames metrics that can not be used, is that correct? We should use a different field, s.Spec.ContainerSpec.Labels

File: discovery/moby/tasks.go [hunk 1]
```diff
@@ -77,7 +77,7 @@ func (d *Discovery) refreshTasks(ctx context.Context) ([]*targetgroup.Group, err
 
 		for k, v := range s.Labels {
```
Comment (roidelapluie) [inline discovery/moby/tasks.go:78]: s.Spec.ContainerSpec.Labels

With nil pointer checks  if needed

File: docs/configuration/configuration.md [hunk 2]
```diff
@@ -795,7 +795,7 @@ Available meta labels:
 * `__meta_dockerswarm_task_id`: the id of the task
 * `__meta_dockerswarm_task_container_id`: the container id of the task
 * `__meta_dockerswarm_task_desired_state`: the desired state of the task
-* `__meta_dockerswarm_task_label_<labelname>`: each label of the task
+* `__meta_dockerswarm_container_label_<labelname>`: each label of the container
```
[the remark this answers is not in the document]
Comment (importhuman) [reply docs/configuration/configuration.md]: fixed

**Comment by roidelapluie** [comment]
Thanks!

**Comment by importhuman** [comment]
@roidelapluie can you please elaborate on what this meant?

> I think the current version renames metrics that can not be used

Thanks for the help with the PR! :smile:

**Comment by importhuman** [comment]
Thanks! Can you suggest another issue to pick up?

**Comment by roidelapluie** [comment]
> Thanks! Can you suggest another issue to pick up?

What would you think about https://github.com/prometheus/prometheus/issues/10200 ?

**Pull request merged**